### PR TITLE
[batch] Remove worker shim classes for cloud credentials

### PR DESCRIPTION
**Pull request by daniel-goldstein**
This change removes shim credential classes from the batch worker code and attempts to consolidate credential handling in the worker API classes.

### CODE REVIEW

**Comment by daniel-goldstein** [comment]
@danking Curious for your thoughts on this refactor. I was getting pretty turned around myself with the various credential classes and I think this is closer to what we want in a keyless world. Ideally the batch worker should just be able to request credentials (in the form of an access token) for a given identity with just the identity's ID.

LMK if you're in favor of this or not, or if you are if you would like to see it folded into the metadata server PR.

File: batch/batch/worker/worker.py [hunk 1]
```diff
@@ -537,9 +536,7 @@ async def pull():
                     await self._ensure_image_is_pulled()
                 elif self.is_public_image:
                     await self._ensure_image_is_pulled(auth=self._batch_worker_registry_credentials)
-                elif self.image_ref_str == BATCH_WORKER_IMAGE and isinstance(
-                    self.credentials, (JVMUserCredentials, CopyStepCredentials)
-                ):
+                elif self.image_ref_str == BATCH_WORKER_IMAGE and self.hail_identity is None:
```
Comment (danking) [inline batch/batch/worker/worker.py:539]: should this say `is not None`?

Comment (danking) [reply batch/batch/worker/worker.py:539]: I'm not sure I follow what the previous if check was checking.

File: batch/batch/worker/worker.py [hunk 2]
```diff
@@ -2323,11 +2320,14 @@ async def run(self):
                     mjs_fut.cancel()
 
     async def cleanup(self):
+        assert CLOUD_WORKER_API
         assert self.worker
         assert self.worker.file_store is not None
         assert self.worker.fs
         assert self.jvm
 
+        await CLOUD_WORKER_API.remove_user_credentials(self.hail_identity)
```
Comment (danking) [inline batch/batch/worker/worker.py:2329]: ~~I'd expect to need some kind of reference counting; why don't we need that? Couldn't multiple jobs share creds?~~

Heh. I think I wouldn't call it remove. Not sure what the right name is but remove feels wrong. Maybe "release"?

**Review by danking — DISMISSED** [review]
Fewer classes => fewer concepts => easier for me to understand

On the other hand, implementing our own GC (the reference counting) seems not great?  I feel like you can integrate the per-cloud stuff from the credentials classes into the WorkerAPI while keeping one-object-with-one-secret-credential-per-job. That avoids the reference counting. I'm not sure I see what the reference counting buys you.

File: batch/batch/worker/worker.py (hunk 1 above)
Comment (daniel-goldstein) [reply batch/batch/worker/worker.py]: I believe my change is consistent, but whether that clause needs to exist at all is a good question. The previous code regarding `JVMUserCredentials` and `CopyStepCredentials` allowed input/output containers and JVM containers, which we determine should use the extant batch worker image, to skip an auth check on the image. I originally changed that field from `Union[CloudUserCredentials, JVMUserCredentials, CopyStepCredentials]` to just `Optional[str]` so the `JVMUserCredentials` and `CopyStepCredentials` case was replaced with `None`.

**Comment by daniel-goldstein** [comment]
Ya you're right, I was over-optimizing trying to share credentials between jobs of the same user. Just kept it as 1:1 jobs to credentials and it got a lot simpler. I changed the key to the job id because using the name of the identity would cause collisions between jobs.

**Review by danking — DISMISSED** [review]
Tell me if I'm bike shedding, but I don't see the value in the register/remove methods. It seems to me that the jobs could just hold the `body['gsa_key']` dicts and then the worker api exposes `container_registry_credentials` and `mount_cloudfuse`.

The win here seems to be reducing the number of distinct concepts/classes by folding it into the WorkerAPI. But having credentials represented as an object which is free'd when the refcount goes to zero seems strictly better than registering and unregistering (which is still effectively a kind of GC, albeit a unique_ptr like one rather than shared_ptr).

**Comment by daniel-goldstein** [comment]
No you're right, I was roping in unrelated changes to this PR. I feel like we need some way of "registering" credentials with the metadata server, but I will leave that to that PR.

**Review by danking — APPROVED** [review]
Beautiful, +70,-161 my favorite kind of PR.

**Comment by danking** [comment]
You've got conflicts though, probably formatting?

**Pull request merged**